### ISSUE TITLE
[@types/jasmine] Jasmine SpyOn type errors when spying on an overload function

### ISSUE THREAD

**Issue by cnishina** [DefinitelyTyped/DefinitelyTyped#39162]
- [X] I tried using the `@types/jasmine` + `@types/node` package and had problems. (Explanation below why I believe it is an `@types/jasmine` issue. The current version I am using is 3.4.4.
- [X] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript = ^3.6.4
- [X] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [X] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @davidparsson @borisyankov

Repro TypeScript example with `@types/node` and `@types/jasmine`:

```
  const existingFiles = [ 'foo.zip', 'foo_.zip' ];
  spyOn(fs, 'readdirSync').and.returnValue(existingFiles);
```

When spying on the value, I am expecting a string[] which is one of the overloads that the function could return; however this defaults to fs.Dirent[]. This used to be okay a few type definitions ago at `@types/jasmine` 3.1.1 this used to work (this is why I think it is a Jasmine issue). The error is:

```
Argument of type 'string[]' is not assignable to parameter of type 'Dirent[]'.
  Type 'string' is not assignable to type 'Dirent'.ts(2345)
```

Here is the fs readdirSync overloads. string[] should be valid as a return value:
```
    /**
     * Synchronous readdir(3) - read a directory.
     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
     * @param options The encoding (or an object specifying the encoding), used as the encoding of the result. If not provided, `'utf8'` is used.
     */
    function readdirSync(path: PathLike, options?: { encoding: BufferEncoding | null; withFileTypes?: false } | BufferEncoding | null): string[];

    /**
     * Synchronous readdir(3) - read a directory.
     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
     * @param options The encoding (or an object specifying the encoding), used as the encoding of the result. If not provided, `'utf8'` is used.
     */
    function readdirSync(path: PathLike, options: { encoding: "buffer"; withFileTypes?: false } | "buffer"): Buffer[];

    /**
     * Synchronous readdir(3) - read a directory.
     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
     * @param options The encoding (or an object specifying the encoding), used as the encoding of the result. If not provided, `'utf8'` is used.
     */
    function readdirSync(path: PathLike, options?: { encoding?: string | null; withFileTypes?: false } | string | null): string[] | Buffer[];

    /**
     * Synchronous readdir(3) - read a directory.
     * @param path A path to a file. If a URL is provided, it must use the `file:` protocol.
     * @param options If called with `withFileTypes: true` the result data will be an array of Dirent.
     */
    function readdirSync(path: PathLike, options: { encoding?: string | null; withFileTypes: true }): Dirent[];
```

Again this used to work and have tested `@types/jasmine` 3.3.1 with both `@types/node` 10.12.12 and 12.11.1. My current workaround is to avoid types in this step. The test passes when I run Jasmine but it is just a type definition issue. See the following: 
```
spyOn(fs, 'readdirSync').and.returnValue(existingFiles as any);
```

**Comment by chivesrs** [comment]
Seems to be similar to #42455 and probably solved by #43004.

**Comment by orta** [comment]
Hi thread, we're moving DefinitelyTyped to use [GitHub Discussions](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/53377) for conversations the `@types` modules in DefinitelyTyped.

To help with the transition, we're closing all issues which haven't had activity in the last 6 months, which includes this issue. If you think closing this issue is a mistake, please pop into the [TypeScript Community Discord](https://discord.gg/typescript) and mention the issue in the `definitely-typed` channel.
